tensorflow_addons/layers/noisy_dense.py

```diff
@@ -0,0 +1,290 @@
+        self.last_dim = tf.compat.dimension_value(input_shape[-1])
+        sqrt_dim = self.last_dim ** (1 / 2)
+        if self.last_dim is None:
+            raise ValueError(
+                "The last dimension of the inputs to `Dense` "
+                "should be defined. Found `None`."
+            )
+        self.input_spec = InputSpec(min_ndim=2, axes={-1: self.last_dim})
+
+        self.sigma_init = initializers.Constant(value=self.sigma / sqrt_dim)
+        self.mu_init = initializers.RandomUniform(
+            minval=-1 / sqrt_dim, maxval=1 / sqrt_dim
+        )
+
+        # Learnable parameters
+        self.sigma_kernel = self.add_weight(
+            "sigma_kernel",
+            shape=[self.last_dim, self.units],
+            initializer=self.sigma_init,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+            dtype=self.dtype,
+            trainable=True,
+        )
+
+        self.mu_kernel = self.add_weight(
+            "mu_kernel",
+            shape=[self.last_dim, self.units],
+            initializer=self.mu_init,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+            dtype=self.dtype,
+            trainable=True,
+        )
+
+        if self.use_bias:
+            self.sigma_bias = self.add_weight(
+                "sigma_bias",
+                shape=[
+                    self.units,
+                ],
+                initializer=self.sigma_init,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+                dtype=self.dtype,
+                trainable=True,
+            )
+
+            self.mu_bias = self.add_weight(
+                "mu_bias",
+                shape=[
+                    self.units,
+                ],
+                initializer=self.mu_init,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+                dtype=self.dtype,
+                trainable=True,
+            )
+        else:
+            self.sigma_bias = None
+            self.mu_bias = None
+        self.built = True
+
+    # Create the factorised Gaussian noise
+    def reset_noise(self):
+        dtype = self._compute_dtype_object
+
+        # Generate random noise
+        eps_i = tf.random.normal([self.last_dim, self.units], dtype=dtype)
+        eps_j = tf.random.normal(
+            [
+                self.units,
+            ],
+            dtype=dtype,
+        )
+
+        # Scale the random noise
+        self.eps_kernel = _scale_noise(eps_i) * _scale_noise(eps_j)
+        self.eps_bias = _scale_noise(eps_j)
+
+    def remove_noise(self):
+        dtype = self._compute_dtype_object
+        self.eps_kernel = tf.zeros([self.last_dim, self.units], dtype=dtype)
+        self.eps_bias = tf.zeros([self.last_dim, self.units], dtype=dtype)
+
+    def call(self, inputs, reset_noise=True, remove_noise=False):
```

Comment:
This is good to me 😃

Reply:
Private python :wink:

Reply:
Not really private :-( Just hide them from doc LOL